nmigen_boards/quickfeather.py

```diff
@@ -0,0 +1,86 @@
+import os
+import sys
+import subprocess
+
+from nmigen.build import *
+from nmigen.vendor.quicklogic import *
+from nmigen_boards.resources import *
+
+
+__all__ = ["QuickfeatherPlatform"]
+
+
+class QuickfeatherPlatform(QuicklogicPlatform):
+    device      = "ql-eos-s3_wlcsp"
+    part        = "PU64"
+    default_clk = "sys_clk0"
+    connectors = [
+        Connector("J", 2, "- 28 22 21 37 36 42 40 7 2 4 5"),
+        Connector("J", 3, "- 8 9 17 16 20 6 55 31 25 47 - - - - 41"),
+        Connector("J", 8, "27 26 33 32 23 57 56 3 64 62 63 61 59 - - -"),
+    ]
+    resources   = [
+        # This is a placeholder resource since the board utilizes
+        # default internal SoC clock.
+        Resource("sys_clk0", 0, Pins("63", dir="i"), Clock(10e6)),
```

Comment:
I believe that a placeholder should not be necessary here. If you are getting exceptions in `default_clk_constraint`, it means you need to implement something similar to what ECP5 does:
```
    @property
    def default_clk_constraint(self):
        if self.default_clk == "sys_clk0":
            return Clock(10e6)
        return super().default_clk_constraint
```
That's assuming `sys_clk0` actually runs at 10 MHz. Is that the case, or is it dynamically configured by the CPU at runtime?

Reply:
The default frequency is set to around 77MHz according to docs, but the autogenerated openocd script changes frequency to 60MHz (I corrected the value, it was wrongly set to 10MHz). The frequency can be configured by writing to the `prog` register in `AIP` peripheral - I could dynamically change the blink rate by writing different values to it.
Should I add a similar to ECP5 property as you suggested or there is a different approach for such case?

Reply:
Thanks for explaining. This is a tricky case--we don't yet have any board or SoC that works *quite* like this. I can't think of any better way to handle it at the moment, so let's keep the hardcoded 60 MHz value for the moment, and deal with dynamic frequency changes later.

Reply:
Actually, looking closer at some of the tools, do I understand it correctly that the `--osc-freq` of `bitstream_to_openocd.py` means that the FPGA clock frequency can be still set through the FPGA tooling? In that case, this is very similar to the way the ECP5 platform handles `OSCG` and `oscg_div`. If you confirm this, I'll go ahead and adjust the Quicklogic platform to work similarly.

Reply:
I checked the docs and the `.openocd` output once more and indeed there is a setting for a divider. In the end the frequency provided to FPGA is: `{--osc-freq} / {--fpga-clk-divider}` (args from `bitstream_to_openocd.py`), so for default values we've got: `60e6 / 12 = 5MHz` actually. I believe we can use the same solution as in ECP5 and if needed we can also enable an oscillator frequency setup with the `--osc-freq` argument.

I've prepared a PR to adapt Quicklogic platform for such usage: [link](https://github.com/nmigen/nmigen/pull/543), please take a look.
I applied your patch and it worked nicely (with and without `do_program=True`) along with ECP5-like solution.